### PR TITLE
Add -AsHashtable switch to ConvertFrom-Json for issue #3623

### PR DESCRIPTION
**Pull request by bergmeister**
Address https://github.com/PowerShell/PowerShell/issues/3623.

Implementation symmetric to traditional code path that returns a `PSCustomObject`. Code is shared on the top level but 2 low level methods were difficult to share, therefore 2 separate but similar methods were created for HashTables.
All existing tests related to this change were adapted to also test against this new switch and were slightly improved. 

Update (@TravisEz13):
`-AsHashtable` is an existing pattern used in `Group-Object`

### CODE REVIEW

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs [hunk 1]
```diff
@@ -26,13 +27,14 @@ public static class JsonObject
...
+        /// Convert a Json string back to an object of type PSObject or Hashtable depending on parameter <paramref name="returnHashTable"/>.
         /// </summary>
         /// <param name="input"></param>
+        /// <param name="returnHashTable"></param>
         /// <param name="error"></param>
-        /// <returns></returns>
+        /// <returns>A PSObject or a Hashtable if the <paramref name="returnHashTable"/> parameter is true.</returns>
         [SuppressMessage("Microsoft.Naming", "CA1704:IdentifiersShouldBeSpelledCorrectly")]
-        public static object ConvertFromJson(string input, out ErrorRecord error)
+        public static object ConvertFromJson(string input, bool returnHashTable, out ErrorRecord error)
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs]: We shouldn't change a public API. Please add new overload.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1 [hunk 2]
```diff
@@ -1,46 +1,33 @@
...
-        @{ name = "empty"; str = "" }
-        @{ name = "spaces"; str = "  " }
-        @{ name = "object"; str = "{a:1}" }
+        @{ name = "empty";  str = "";      ReturnHashTable = $true  }
+        @{ name = "spaces"; str = "  ";    ReturnHashTable = $true  }
+        @{ name = "object"; str = "{a:1}"; ReturnHashTable = $true  }
+        @{ name = "empty";  str = "";      ReturnHashTable = $false }
+        @{ name = "spaces"; str = "  ";    ReturnHashTable = $false }
+        @{ name = "object"; str = "{a:1}"; ReturnHashTable = $false }
     )
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1]: Please move to BeforeAll block.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1 [hunk 3]
```diff
@@ -1,46 +1,33 @@
...
 
-    It 'no error for valid string - <name>' -TestCase $validStrings {
-        param ($str)
+    It 'no error for valid string - <name> when ReturnHashTable is <ReturnHashTable>' -TestCase $validStrings {
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1]: Maybe 'no error for valid string - <name> with -ReturnHashTable:`$\<ReturnHashTable>'

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/ConvertFromJsonCommand.cs [hunk 4]
```diff
@@ -20,17 +20,23 @@ public class ConvertFromJsonCommand : Cmdlet
...
         /// <summary>
-        /// inputObjectBuffer buffers all InputObjet contents available in the pipeline.
+        /// InputObjectBuffer buffers all InputObject contents available in the pipeline.
         /// </summary>
         private List<string> _inputObjectBuffer = new List<string>();
 
+        /// <summary>
+        /// Returned data structure is a Hashtable instead a CustomPSObject.
+        /// </summary>
+        [Parameter(Mandatory = false)]
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/ConvertFromJsonCommand.cs]: It is default. Please remove.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs [hunk 5]
```diff
@@ -65,15 +67,29 @@ public static object ConvertFromJson(string input, out ErrorRecord error)
...
-                        obj = PopulateFromJArray(list, out error);
+                        if (returnHashTable)
+                        {
+                            obj = PopulateHashTableFromJArray(list, out error);
+                        }
+                        else
+                        {
+                            obj = PopulateFromJArray(list, out error);
+                        }
                     }
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs]: Please add 'else' with Diagnostics.Assert (if the object is not JObject or JArray).

**Review by iSazonov — COMMENTED** [review]
@bergmeister Thanks for great contribution!

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs (hunk 1 above)
Comment (bergmeister) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs]: Thanks. I did not know PowerShell also exposes some of its classes. I'll re-add the old signature API  that will then internally call the new overload.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs (hunk 5 above)
Comment (bergmeister) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs]: Because of the cast to `JArray` a few lines above, `obj` cannot be something else. I initially put your suggestion in but even the compiler tells me that:
`error CS0184: The given expression is never of the provided ('JObject') type`

File: test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1 (hunk 2 above)
Comment (bergmeister) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1]: I am not sure what you exactly mean by that. I removed the redundant line about `$TestCasesForReturnHashTableParameter that was a leftover.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/ConvertFromJsonCommand.cs (hunk 4 above)
Comment (bergmeister) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/ConvertFromJsonCommand.cs]: What do you mean by that? Can you elaborate please (what is 'it')?

File: test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1 (hunk 3 above)
Comment (bergmeister) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1]: I changed to match your suggestion whilst still displaying the name of the string.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/ConvertFromJsonCommand.cs (hunk 4 above)
Comment (TravisEz13) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/ConvertFromJsonCommand.cs]: `[Parameter(Mandatory = false)]` is default

Comment (bergmeister) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/ConvertFromJsonCommand.cs]: Ok. I see, thanks.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs (hunk 5 above)
Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs]: I meant:
```c#
                        Dbg.Diagnostics.Assert(
                            false,
                            "Only JObject or JArray is supported.");

```

File: test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1 (hunk 2 above)
Comment (iSazonov) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1]: Please see [sample](https://github.com/PowerShell/PowerShell/blob/309bf603f9aff9009dce7e725d42b2d4f99f58c9/test/powershell/Modules/Microsoft.PowerShell.Management/Clear-Item.Tests.ps1#L2) 
We should place ` $TestCasesForReturnHashTableParameter ` and `$validStrings ` in the BeforeAll block.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/InvokeRestMethodCommand.Common.cs [hunk 6]
```diff
@@ -171,7 +171,7 @@ private bool TryConvertToJson(string json, out object obj, ref Exception exRef)
             try
             {
                 ErrorRecord error;
-                obj = JsonObject.ConvertFromJson(json, out error);
+                obj = JsonObject.ConvertFromJson(json, false, out error);
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/InvokeRestMethodCommand.Common.cs]: Please revert the change - we have the overload.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1 [hunk 7]
```diff
@@ -1,46 +1,31 @@
...
         $errRecord | Should BeNullOrEmpty
     }
 
     $invalidStrings = @(
-        @{ name = "plain text"; str = "plaintext" }
-        @{ name = "part"; str = '{"a" :' }
+        @{ name = "plain text"; str = "plaintext"; ReturnHashTable = $true  }
+        @{ name = "part";       str = '{"a" :';    ReturnHashTable = $true  }
+        @{ name = "plain text"; str = "plaintext"; ReturnHashTable = $false }
+        @{ name = "part";       str = '{"a" :';    ReturnHashTable = $false }
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1]: I see we have a check for `input == null` but haven't the test - please add new test with `str = $null`.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/InvokeRestMethodCommand.Common.cs (hunk 6 above)
Comment (bergmeister) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/InvokeRestMethodCommand.Common.cs]: Yes, you're right.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1 (hunk 7 above)
Comment (bergmeister) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1]: OK. Added

**Comment by TravisEz13** [comment]
`-AsHashtable` is an existing pattern used in `Group-Object`

**Comment by iSazonov** [comment]
We have discussion for this https://github.com/PowerShell/PowerShell/issues/3623

**Comment by TravisEz13** [comment]
@dantraMSFT Can you review this?

**Comment by SteveL-MSFT** [comment]
As part of this PR, if we can detect that the json has an property named empty string for the pscustomobject path and provide an error message like:

```markdown
The provided json includes a property who name is an empty string, this is only supported using the `-AsHashTable` switch.
```

Then we can resolve https://github.com/PowerShell/PowerShell/issues/1755 as fixed.

File: src/Microsoft.PowerShell.Commands.Utility/resources/WebCmdletStrings.resx [hunk 8]
```diff
@@ -126,9 +126,12 @@
   <data name="CredentialConflict" xml:space="preserve">
     <value>The cmdlet cannot run because the following conflicting parameters are specified: Credential and UseDefaultCredentials. Specify either Credential or UseDefaultCredentials, then retry.</value>
   </data>
-  <data name="DirecotryPathSpecified" xml:space="preserve">
+  <data name="DirectoryPathSpecified" xml:space="preserve">
     <value>Path '{0}' resolves to a directory. Specify a path including a file name, and then retry the command.</value>
   </data>
+  <data name="EmptyKeyInJsonString" xml:space="preserve">
+    <value>The provided json includes a property whose name is an empty string, this is only supported using the -AsHashTable switch.</value>
```
Comment (SteveL-MSFT) [inline src/Microsoft.PowerShell.Commands.Utility/resources/WebCmdletStrings.resx]: Should probably capitalize json as `JSON`

**Review by SteveL-MSFT — CHANGES_REQUESTED** [review]

File: src/Microsoft.PowerShell.Commands.Utility/resources/WebCmdletStrings.resx (hunk 8 above)
Comment (bergmeister) [reply src/Microsoft.PowerShell.Commands.Utility/resources/WebCmdletStrings.resx]: OK. Fixed it.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1 [hunk 9]
```diff
@@ -1,46 +1,47 @@
...
+        @{ name = "object"; str = "{a:1}"; ReturnHashTable = $true  }
+        @{ name = "null";   str = $null;   ReturnHashTable = $false }
+        @{ name = "empty";  str = "";      ReturnHashTable = $false }
+        @{ name = "spaces"; str = "  ";    ReturnHashTable = $false }
+        @{ name = "object"; str = "{a:1}"; ReturnHashTable = $false }
     )
 
-    It 'no error for valid string - <name>' -TestCase $validStrings {
-        param ($str)
+    It 'no error for valid string ''<name>'' with -ReturnHashTable:$<ReturnHashTable>' -TestCase $validStrings {
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1]: Please start test descriptions with capital.

We use `$validStrings` one time. Please use our common template:
```powershell
It 'no error for valid string ''<name>'' with -ReturnHashTable:$<ReturnHashTable>' -TestCase @(
 @{ name = "null";   str = $null;   ReturnHashTable = $true  }
 ...
) {
...
}
```

File: test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1 [hunk 10]
```diff
@@ -1,46 +1,47 @@
...
-        @{ name = "part"; str = '{"a" :' }
+        @{ name = "plain text"; str = "plaintext"; ReturnHashTable = $true  }
+        @{ name = "part";       str = '{"a" :';    ReturnHashTable = $true  }
+        @{ name = "plain text"; str = "plaintext"; ReturnHashTable = $false }
+        @{ name = "part";       str = '{"a" :';    ReturnHashTable = $false }
     )
 
-    It 'throw ArgumentException for invalid string - <name>' -TestCase $invalidStrings  {
-        param ($str)
+    It 'throw ArgumentException for invalid string ''<name>'' with -ReturnHashTable:$<ReturnHashTable>' -TestCase $invalidStrings  {
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1]: The same.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1 [hunk 11]
```diff
@@ -1,46 +1,47 @@
...
+        param ($str, $ReturnHashTable)
         $errRecord = $null
-        { [Microsoft.PowerShell.Commands.JsonObject]::ConvertFromJson($str, [ref]$errRecord) } | ShouldThrow 'ArgumentException'
+        { [Microsoft.PowerShell.Commands.JsonObject]::ConvertFromJson($str, $ReturnHashTable, [ref]$errRecord) } | ShouldBeErrorId "ArgumentException"
+    }
+
+    $jsonWithEmptyKey = '{"": "Value"}'
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1]: We should put variables in `BeforeAll` block or in `It` block.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1 [hunk 12]
```diff
@@ -1,46 +1,47 @@
...
+    It 'throw InvalidOperationException when json contains empty key name' {
+        $errorRecord = $null
+        [Microsoft.PowerShell.Commands.JsonObject]::ConvertFromJson($jsonWithEmptyKey, [ref]$errorRecord)
+        $errorRecord.Exception.GetType() | Should Be System.InvalidOperationException
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1]: Usually we don't check an exception types. Please remove.
(For type check we use Should BeOfType)

File: test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1 [hunk 13]
```diff
@@ -1,46 +1,47 @@
...
+        $errorRecord.FullyQualifiedErrorId | Should Be EmptyKeyInJsonString
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1]: Please put `EmptyKeyInJsonString` in quotas. It is more readable.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-Json.Tests.ps1 [hunk 14]
```diff
@@ -1,16 +1,33 @@
 Describe 'ConvertFrom-Json' -tags "CI" {
-    It 'can convert a single-line object' {
-        ('{"a" : "1"}' | ConvertFrom-Json).a | Should Be 1
+    
+    $testCasesWithAndWithoutAsHashtableSwitch = @(
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-Json.Tests.ps1]: Please put this in `BeforeAll` block.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-Json.Tests.ps1 [hunk 15]
```diff
@@ -1,16 +1,33 @@
...
+        @{ AsHashtable = $true  }
+        @{ AsHashtable = $false }
+    )
+
+    It 'can convert a single-line object with AsHashtable switch set to <AsHashtable>' -TestCase $testCasesWithAndWithoutAsHashtableSwitch {
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-Json.Tests.ps1]: Please start test descriptions with capital.

**Review by iSazonov — COMMENTED** [review]
Do we address the [comment](https://github.com/PowerShell/PowerShell/issues/3623#issuecomment-296537316) about case sensitivity?

File: test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1 [hunk 16]
```diff
@@ -1,6 +1,11 @@
 ﻿Describe 'Unit tests for JsonObject' -tags "CI" {
 
-    $validStrings = @(
+    BeforeAll {
+        $jsonWithEmptyKey = '{"": "Value"}'
+    }
+
+
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1]: Please remove extra line.

**Comment by iSazonov** [comment]
@SteveL-MSFT @TravisEz13 @dantraMSFT Could you please review?

Also should we address the [comment](https://github.com/PowerShell/PowerShell/issues/3623#issuecomment-296537316) from @powercode 
>Apart from object size and performance, it may matter with case sensitivity. I have run into issues where I couldn't use the PowerShell cmdlets because the json contained 'Property' and 'property'.

**Comment by iSazonov** [comment]
@bergmeister Thanks for your contribution! LGTM with above question about  case sensitivity.

**Comment by SteveL-MSFT** [comment]
@iSazonov @powercode regarding case sensitivity, I think we should treat it the same as the empty key.  [Hashtable] allows two keys differing in case, so if we detect that, we should just tell the user to use `-AsHashTable`.

**Comment by bergmeister** [comment]
@SteveL-MSFT @iSazonov @powercode I added a specific error message to point the user to the new `-AsHashTable` option.
As part of this I also found out that actually no error is thrown in the case of duplicate keys with the same casing. This is due to the behaviour of Newtonsoft.Json that just uses the last item if there are multiple items with the same key of the same casing. The PowerShell code currently caters for it should this behaviour change in a newer version of Newtonsoft.Json. This should definitely be discussed in another issue if you want to make a breaking change to throw when encountering case matching duplicate keys.

File: src/Microsoft.PowerShell.Commands.Utility/resources/WebCmdletStrings.resx [hunk 17]
```diff
@@ -133,7 +133,10 @@
     <value>The provided JSON includes a property whose name is an empty string, this is only supported using the -AsHashTable switch.</value>
   </data>
   <data name="DuplicateKeysInJsonString" xml:space="preserve">
-    <value>Cannot convert the JSON string because a dictionary that was converted from the string contains the duplicated keys '{0}' and '{1}'.</value>
+    <value>Cannot convert the JSON string because a dictionary that was converted from the string contains the duplicated key '{0}'.</value>
+  </data>
+  <data name="KeysWithDifferentCasingInJsonString" xml:space="preserve">
+    <value>Cannot convert the JSON string because a PSObject does not support keys with different casing. Please use the -AsHashTable switch instead. The key that was attempted to be added to the exisiting key '{0}' was '{1}'.</value>
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/resources/WebCmdletStrings.resx]: Oh, very long message!

File: test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1 [hunk 18]
```diff
@@ -1,46 +1,63 @@
...
+        It 'Throw InvalidOperationException when json contains key with different casing' {
+            $errorRecord = $null
+            [Microsoft.PowerShell.Commands.JsonObject]::ConvertFromJson($jsonContainingKeysWithDifferentCasing, [ref]$errorRecord)
+            $errorRecord.FullyQualifiedErrorId | Should Be 'KeysWithDifferentCasingInJsonString'
+        }
+
+        It 'Not throw when json contains key (same casing) when ReturnHashTable is true' {
+            $errorRecord = $null
+            $result = [Microsoft.PowerShell.Commands.JsonObject]::ConvertFromJson($jsonContainingKeysWithDifferentCasing, $true, [ref]$errorRecord)
+            $result | Should Not Be $null
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1]: I think it makes sense to check the hashtable values.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1 [hunk 19]
```diff
@@ -1,46 +1,63 @@
...
+        @{ name = "object"; str = "{a:1}"; ReturnHashTable = $true  }
+        @{ name = "null";   str = $null;   ReturnHashTable = $false }
+        @{ name = "empty";  str = "";      ReturnHashTable = $false }
+        @{ name = "spaces"; str = "  ";    ReturnHashTable = $false }
+        @{ name = "object"; str = "{a:1}"; ReturnHashTable = $false }
+    ) {
+        param ($str, $ReturnHashTable)
         $errRecord = $null
-        [Microsoft.PowerShell.Commands.JsonObject]::ConvertFromJson($str, [ref]$errRecord)
+        [Microsoft.PowerShell.Commands.JsonObject]::ConvertFromJson($str, $ReturnHashTable, [ref]$errRecord)
```
Comment (SteveL-MSFT) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1]: Please validate the returned object

File: test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1 [hunk 20]
```diff
@@ -1,46 +1,63 @@
...
+        It 'Throw InvalidOperationException when json contains empty key name' {
+            $errorRecord = $null
+            [Microsoft.PowerShell.Commands.JsonObject]::ConvertFromJson($jsonWithEmptyKey, [ref]$errorRecord)
+            $errorRecord.FullyQualifiedErrorId | Should Be 'EmptyKeyInJsonString'
+        }
+
+        It 'Not throw when json contains empty key name when ReturnHashTable is true' {
+            $errorRecord = $null
+            $result = [Microsoft.PowerShell.Commands.JsonObject]::ConvertFromJson($jsonWithEmptyKey, $true, [ref]$errorRecord)
+            $result | Should Not Be $null
```
Comment (SteveL-MSFT) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1]: Please validate the returned object

File: test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1 (hunk 18 above)
Comment (SteveL-MSFT) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1]: Please validate the returned object

**Review by iSazonov — COMMENTED** [review]
@bergmeister Thanks for your contribution! I believe we are near to merge.
Feel free to open new Issue to discuss the discovered problem.

**Review by SteveL-MSFT — CHANGES_REQUESTED** [review]

File: src/Microsoft.PowerShell.Commands.Utility/resources/WebCmdletStrings.resx (hunk 17 above)
Comment (bergmeister) [reply src/Microsoft.PowerShell.Commands.Utility/resources/WebCmdletStrings.resx]: I know. I shortened it a bit but I think the last sentence still adds value.

**Comment by bergmeister** [comment]
@iSazonov @SteveL-MSFT I added more validation to the positive test cases and shortened the error message a bit. I opened issue #5199 for the problem with Newtonsoft.Json for keys with matching cases, which is unrelated to this PR.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1 [hunk 21]
```diff
@@ -1,46 +1,68 @@
...
+            $result.Count | Should Be 1
+            $result.$('') | Should Be 'Value'
```
Comment (SteveL-MSFT) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1]: You should simplify this to just: $result."" | Should Be 'Value'

File: test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1 (hunk 19 above)
Comment (SteveL-MSFT) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1]: You should capture the output as $result and validate it is what you expect

File: test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1 (hunk 18 above)
Comment (SteveL-MSFT) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1]: This one is resolved.  Thanks.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1 (hunk 20 above)
Comment (SteveL-MSFT) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1]: Resolved

**Review by iSazonov — APPROVED** [review]

**Review by SteveL-MSFT — APPROVED** [review]

**Comment by bergmeister** [comment]
@SteveL-MSFT I see that the documentation [here](https://github.com/PowerShell/PowerShell-Docs/blob/a1520688fb6b3e25ec51096d2611de9b54d08c2d/reference/6/Microsoft.PowerShell.Utility/ConvertFrom-Json.md) has not been updated yet for the new `-AsHashTable` switch (or is it already done in another branch?). Am I supposed to do that or shall I simply open an issue in the docs repo?

**Comment by iSazonov** [comment]
@bergmeister Feel free to open an Issue in docs repo or push PR with the fix.

**Comment by SteveL-MSFT** [comment]
@bergmeister preferably submit a PR to powershell-docs repo.  The `Documentation Needed` label here means you don't need to open an issue in the powershell-docs repo.  But my team won't be reviewing and adding docs until sometime after RC is out.  If you submit a PR for the doc update, please reference this PR so we can remove the `Documentation Needed` label.  Thanks!
